src/control/lib/hostlist/hostlist.go

```diff
@@ -0,0 +1,735 @@
+	defer hl.RUnlock()
+
+	var bld strings.Builder
+
+	for i, hr := range hl.ranges {
+		bld.WriteString(hr.derangedString())
+		if i < len(hl.ranges)-1 {
+			bld.WriteString(",")
+		}
+	}
+
+	return bld.String()
+}
+
+// Push adds a string representation of hostnames to this HostList.
+func (hl *HostList) Push(stringHosts string) error {
+	other, err := Create(stringHosts)
+	if err != nil {
+		return err
+	}
+
+	return hl.PushList(other)
+}
+
+// PushHost adds a single host to this HostList.
+func (hl *HostList) PushHost(stringHost string) error {
+	hn, err := createHostName(stringHost)
+	if err != nil {
+		return err
+	}
+
+	hl.Lock()
+	defer hl.Unlock()
+
+	if hn.hasNumber {
+		return hl.pushRange(&hostRange{
+			prefix:  hn.prefix,
+			suffix:  hn.suffix,
+			lo:      hn.number,
+			hi:      hn.number,
+			width:   hn.width,
+			isRange: true,
+		})
+	}
+
+	return hl.pushRange(&hostRange{
+		prefix: stringHost,
+	})
+}
+
+func (hl *HostList) pushRange(hr *hostRange) error {
+	if len(hl.ranges) > 0 {
+		tail := hl.ranges[len(hl.ranges)-1]
+		if tail.canAppend(hr) {
+			tail.hi = hr.hi
+			hl.hostCount += hr.count()
+			return nil
+		}
+	}
+
+	hl.ranges = append(hl.ranges, hr)
+	hl.hostCount += hr.count()
+
+	return nil
+}
+
+// PushList adds the supplied HostList onto this HostList.
+func (hl *HostList) PushList(other *HostList) error {
+	hl.Lock()
+	defer hl.Unlock()
+	other.RLock()
+	defer other.RUnlock()
+
+	for _, hr := range other.ranges {
+		if err := hl.pushRange(hr); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func fmtRangeHost(hr *hostRange, num uint) string {
+	return fmt.Sprintf("%s%0*d%s", hr.prefix, hr.width, num, hr.suffix)
+}
+
+// Pop returns the string representation of the last host pushed
+// onto the HostList and then removes it from the HostList. Returns
+// an error if the HostList is empty.
+func (hl *HostList) Pop() (hostName string, err error) {
+	if hl.IsEmpty() {
+		return "", ErrEmpty
+	}
+
+	hl.Lock()
+	defer hl.Unlock()
+
+	hl.hostCount--
+
+	tailIdx := len(hl.ranges) - 1
+	tail := hl.ranges[tailIdx]
+	if !tail.isRange {
+		hostName = tail.prefix
+		hl.ranges = hl.ranges[:tailIdx]
+		return
+	}
+
+	hostName = fmtRangeHost(tail, tail.hi)
+	tail.hi--
+	if tail.hi < tail.lo {
+		hl.ranges = hl.ranges[:tailIdx]
+	}
+
+	return
+}
+
+// Shift returns the string representation of the first host pushed
+// onto the HostList and then removes it from the HostList. Returns
+// an error if the HostList is empty.
+func (hl *HostList) Shift() (hostName string, err error) {
+	if hl.IsEmpty() {
+		return "", ErrEmpty
+	}
+
+	hl.Lock()
+	defer hl.Unlock()
+
+	hl.hostCount--
+
+	head := hl.ranges[0]
+	if !head.isRange {
+		hostName = head.prefix
+		hl.ranges = hl.ranges[1:]
+		return
+	}
+
+	hostName = fmtRangeHost(head, head.lo)
+	head.lo++
+	if head.lo > head.hi {
+		hl.ranges = hl.ranges[1:]
+	}
+
+	return
+}
+
+func (hl *HostList) getNthHostRange(n int) (int, *hostRange, uint, error) {
+	if hl.hostCount == 0 {
```

Comment:
Do you need locks here?  Seems like the data has potential to change.

Reply:
These are not public methods, so they can't be called from outside the package. The public methods are responsible for managing the locks before calling these internal methods.

Reply:
Good point.  Thanks for the clarification.